Libraries/LibCrypto/NumberTheory/ModularFunctions.h

```diff
@@ -0,0 +1,202 @@
+#endif
+        if (ep.words()[0] % 2 == 1) {
+            exp = exp.multiplied_by(base).divided_by(m).remainder;
+        }
+        ep = ep.divided_by(2).quotient;
+        base = base.multiplied_by(base).divided_by(m).remainder;
+    }
+    return exp;
+}
+
+static auto GCD(const UnsignedBigInteger& a, const UnsignedBigInteger& b) -> UnsignedBigInteger
+{
+    UnsignedBigInteger a_ { a }, b_ { b };
+    for (;;) {
+        if (a_ == 0)
+            return b_;
+        b_ = b_.divided_by(a_).remainder;
+        if (b_ == 0)
+            return a_;
+        a_ = a_.divided_by(b_).remainder;
+    }
+}
+
+static auto LCM(const UnsignedBigInteger& a, const UnsignedBigInteger& b) -> UnsignedBigInteger
+{
+    auto temp = GCD(a, b);
+
+    auto div = a.divided_by(temp);
+
+#ifdef NT_DEBUG
+    dbg() << "quot: " << div.quotient << " rem: " << div.remainder;
+#endif
+    return temp == 0 ? 0 : (a.divided_by(temp).quotient.multiplied_by(b));
+}
+
+template<size_t test_count>
+static bool MR_primality_test(UnsignedBigInteger n, const Vector<UnsignedBigInteger, test_count>& tests)
+{
+    auto prev = n.minus({ 1 });
+    auto b = prev;
+    auto r = 0;
+
+    auto div_result = b.divided_by(2);
+    while (div_result.quotient == 0) {
+        div_result = b.divided_by(2);
+        b = div_result.quotient;
+        ++r;
+    }
+
+    for (size_t i = 0; i < tests.size(); ++i) {
+        auto return_ = true;
+        if (n < tests[i])
+            continue;
+        auto x = ModularPower(tests[i], b, n);
+        if (x == 1 || x == prev)
+            continue;
+        for (auto d = r - 1; d != 0; --d) {
+            x = ModularPower(x, 2, n);
+            if (x == 1)
+                return false;
+            if (x == prev) {
+                return_ = false;
+                break;
+            }
+        }
+        if (return_)
+            return false;
+    }
+
+    return true;
+}
+
+static UnsignedBigInteger random_number(const UnsignedBigInteger& min, const UnsignedBigInteger& max)
+{
+    auto range = max.minus(min);
+    UnsignedBigInteger base;
+    // FIXME: Need a cryptographically secure rng
+    auto size = range.trimmed_length() * sizeof(u32);
+    u8 buf[size];
+    arc4random_buf(buf, size);
+    Vector<u32> vec;
+    for (size_t i = 0; i < size / sizeof(u32); ++i) {
+        vec.append(*(u32*)buf + i);
+    }
+    UnsignedBigInteger offset { move(vec) };
+    return offset.plus(min);
+}
+
+static bool is_probably_prime(const UnsignedBigInteger& p)
+{
+    if (p == 2 || p == 3 || p == 5)
+        return true;
+    if (p < 49)
+        return true;
+
+    Vector<UnsignedBigInteger, 256> tests;
+    UnsignedBigInteger seven { 7 };
+    for (size_t i = 0; i < tests.size(); ++i)
+        tests.append(random_number(seven, p.minus(2)));
+
+    return MR_primality_test(p, tests);
+}
+
+static UnsignedBigInteger random_big_prime(size_t bits)
+{
+    UnsignedBigInteger min = UnsignedBigInteger::from_base10("6074001000").shift_left(bits - 33);
```

Comment:
Same here, let's add an `ASSERT(bits >= 33)` right before ?
If `bits` is less than 33, I think this would wrap around and we would end up with a lot of used memory, as well as `random_number(<big number>, <small number>)` which doesn't seem handled either.

Right now, it's only used when creating new RSA pairs and we never use them with less than 128 bits, so we won't hit that case.